src/convert.jl

```diff
@@ -0,0 +1,230 @@
+    isnothing(idx) && error("Coulnd't find Int type for value $(x)")
+    T = candidates[idx]
+
+    T, TYPE_TO_AUX_INT_FIELDS[T]
+end
+
+function encode_aux_field(record::SAM.Record, field::UnitRange{Int64})
+    typ = record.data[first(field)+3]
+    key = record.data[first(field):first(field)+1]
+
+    if typ == UInt8('i')
+        val = record.data[first(field)+5:last(field)] |> String
+        val = parse(Int32, val)
+        T, typ = find_smallest_int_type(val)
+        val = reinterpret(UInt8, [T(val)])
+
+        return vcat(key, typ, val)
+    elseif typ == UInt8('Z')
+        val = record.data[first(field)+5:last(field)]
+        return vcat(key, typ, val, 0x00)
+    elseif typ == UInt8('A')
+        val = record.data[first(field)+5:last(field)]
+        return vcat(key, typ, val)
+    else
+        error("Unsupported tag type $(Char(typ))")
+    end
+end
+
+function encode_cigar(record::SAM.Record)
+
+    ops = parse_cigar(record)
+
+    n_cigar_op = length(ops)
+    if n_cigar_op > typemax(UInt16) 
+        error("Cigar contains more than $(typemax(UInt16)) operations. This is currently unsupported. See section 4.2.2 of SAM/BAM specifications.")
+    end
+
+    cigar = Vector{UInt8}(undef, length(ops)*4)
+    k = 1
+    for c in ops
+        cigar_32 = [UInt32(0) + OP_TO_CODE[c.op] + (UInt32(c.L) << 4)]
+        for u in reinterpret(UInt8, cigar_32)
+            cigar[k] = u
+            k+=1
+        end
+    end
+    cigar, n_cigar_op
+end
+
+@inline function encode_nucleotide(nt::DNA)
+    if !isvalid(nt)
+        throw(EncodeError(DNAAlphabet{4}(), nt))
+    end
+    return reinterpret(UInt8, nt)
+end
+
+function encode_sequence(record::SAM.Record)
+    !SAM.hassequence(record) && return UInt8[]
+
+    Nsam = SAM.seqlength(record)
+    Nbam = cld(Nsam, 2)
+
+    # Specs : "When l seq is odd the bottom 4 bits of the last byte are undefined, but we recommend writing these as zero."
+    bamseq = fill(0x00, Nbam)
+    samseq = SAM.sequence(record)
+
+    # SAM
+    # 1 2 3 4 5
+    # A T G G T
+    
+    # BAM
+    # 1  2  3
+    # AT GG T0
+
+    for (i, sym) in enumerate(samseq)
+        val = encode_nucleotide(sym)
+        bamseq[div(i + 1, 2)] |= val << (4 * isodd(i))
+    end
+
+    bamseq
+end
+
+function encode_tempname(record::SAM.Record)
+    tempname = SAM.tempname(record)
+    vcat(unsafe_wrap(Vector{UInt8}, tempname), 0x00) #add NULL terminator
+end
+
+""" See section 5.3 of specifications."""
+function reg2bin(start::Int, stop::Int)
+    stop -= 1
+    (start >> 14) == (stop >> 14) && return ((1 << 15) - 1) ÷ 7 + (start >> 14)
+    (start >> 17) == (stop >> 17) && return ((1 << 12) - 1) ÷ 7 + (start >> 17)
+    (start >> 20) == (stop >> 20) && return ((1 << 9)  - 1) ÷ 7 + (start >> 20)
+    (start >> 23) == (stop >> 23) && return ((1 << 6)  - 1) ÷ 7 + (start >> 23)
+    (start >> 26) == (stop >> 26) && return ((1 << 3)  - 1) ÷ 7 + (start >> 26)
+    return 0
+end
```

Comment:
Since XAM depends on Indexes and both are based on the SAM specs it would makes sense to use the one in Indexes. But there's a difference in semantics, the one in Indexes, `reg2bins` returns an array on bins while `reg2bin` returns a single bin. I don't understand this enough to make a decision (the code of reg2bin is straight copy-paste from the specs).

Reply:
Apologies, `reg2bin` is certainly needed here for the conversion. The other `reg2bins` is not required for the conversion.